websockets-jsr/src/main/java/io/undertow/websockets/jsr/WebSocketSessionRemoteEndpoint.java

```diff
@@ -268,57 +268,78 @@ public void sendBinary(final ByteBuffer data) throws IOException {
             data.clear(); //for some reason the TCK expects this, might as well just match the RI behaviour
         }
 
-        @Override
+       @Override
         public void sendText(final String partialMessage, final boolean isLast) throws IOException {
-            if(partialMessage == null) {
+            if (partialMessage == null) {
                 throw JsrWebSocketMessages.MESSAGES.messageInNull();
             }
-            if (binaryFrameSender != null) {
-                throw JsrWebSocketMessages.MESSAGES.cannotSendInMiddleOfFragmentedMessage();
-            }
-            if (textFrameSender == null) {
-                textFrameSender = undertowSession.getWebSocketChannel().send(WebSocketFrameType.TEXT);
-            }
+
+            StreamSinkFrameChannel sender = getTextFrameSender();
+
             try {
-                Channels.writeBlocking(textFrameSender, WebSocketUtils.fromUtf8String(partialMessage));
-                if(isLast) {
-                    textFrameSender.shutdownWrites();
+                Channels.writeBlocking(sender, WebSocketUtils.fromUtf8String(partialMessage));
+                if (isLast) {
+                    sender.shutdownWrites();
                 }
-                Channels.flushBlocking(textFrameSender);
+                Channels.flushBlocking(sender);
             } finally {
                 if (isLast) {
-                    textFrameSender = null;
+                    clearTextFrameSender();
                 }
             }
-
         }
 
         @Override
         public void sendBinary(final ByteBuffer partialByte, final boolean isLast) throws IOException {
-
             if(partialByte == null) {
                 throw JsrWebSocketMessages.MESSAGES.messageInNull();
             }
-            if (textFrameSender != null) {
-                throw JsrWebSocketMessages.MESSAGES.cannotSendInMiddleOfFragmentedMessage();
-            }
-            if (binaryFrameSender == null) {
-                binaryFrameSender = undertowSession.getWebSocketChannel().send(WebSocketFrameType.BINARY);
-            }
+
+            StreamSinkFrameChannel sender = getBinaryFrameSender();
+
             try {
-                Channels.writeBlocking(binaryFrameSender, partialByte);
-                if(isLast) {
-                    binaryFrameSender.shutdownWrites();
+                Channels.writeBlocking(sender, partialByte);
+                if (isLast) {
+                    sender.shutdownWrites();
                 }
-                Channels.flushBlocking(binaryFrameSender);
-            } finally {
+                Channels.flushBlocking(sender);
+            }
+            finally {
                 if (isLast) {
-                    binaryFrameSender = null;
+                    clearBinaryFrameSender();
                 }
             }
             partialByte.clear();
         }
 
+        private synchronized StreamSinkFrameChannel getTextFrameSender() throws IOException {
```

Comment:
Two different senders are guarded by single instance lock. Is there a reason for such arrangement?

Reply:
Thank you for your question. 
The `textFrameSender` and `binaryFrameSender` class variables are initialized and set to `null` during partial message transmission. 
During this process, all other `send()` methods are blocked from sending messages. Specifically, while sending **partial messages** using `textFrameSender`, transmissions using `binaryFrameSender` are also **restricted**.

If we were to apply separate locks for `textFrameSender` and `binaryFrameSender`, there would be a high risk of deadlocks, especially when different methods interact with these variables concurrently. 
To mitigate this, we decided to use a single instance lock to manage both senders.

Additionally, other `send()` methods call `assertNotInFragment()` to check if a partial message is being transmitted. 
Using separate locks in this scenario could similarly lead to deadlocks. 
For these reasons, we opted for a single lock to ensure safe and consistent behavior.

Reply:
@baranowb 
Alternatively, it seems possible to control concurrency by declaring `textFrameSender` and `binaryFrameSender` as `AtomicReference<StreamSinkFrameChannel>` without using `synchronized`.